core/controllers/base_test.py

```diff
@@ -628,16 +583,113 @@ def test_unauthorized_user_exception_raised_when_session_is_invalid(self):
+        self.exit_stack.enter_context(self.super_admin_context())
+        destroy_auth_session_call_counter = self.exit_stack.enter_context(
+            self.swap_with_call_counter(auth_services, 'destroy_auth_session'))
+
+        response = self.get_html_response('/community-library')
+
+        self.assertIn('<library-page>', response.body)
+        self.assertNotIn('site is temporarily unavailable', response.body)
+        self.assertEqual(destroy_auth_session_call_counter.times_called, 0)
+
+    def test_json_response_is_rejected(self):
+        destroy_auth_session_call_counter = self.exit_stack.enter_context(
+            self.swap_with_call_counter(auth_services, 'destroy_auth_session'))
+
+        response = self.get_json('/url_handler', expected_status_int=503)
+
+        self.assertIn('error', response)
+        self.assertEqual(
+            response['error'],
+            'Oppia is currently being upgraded, and the site should be up '
+            'and running again in a few hours. Thanks for your patience!')
+        self.assertEqual(destroy_auth_session_call_counter.times_called, 1)
+
+    def test_json_response_is_not_rejected_when_user_is_super_admin(self):
+        self.exit_stack.enter_context(self.super_admin_context())
+        destroy_auth_session_call_counter = self.exit_stack.enter_context(
+            self.swap_with_call_counter(auth_services, 'destroy_auth_session'))
+
+        response = self.get_json('/url_handler')
+
+        self.assertIn('login_url', response)
+        self.assertIsNone(response['login_url'])
+        self.assertEqual(destroy_auth_session_call_counter.times_called, 0)
+
+    def test_csrfhandler_handler_is_not_rejected(self):
+        response = self.get_json('/csrfhandler')
+
+        self.assertTrue(
+            base.CsrfTokenManager.is_csrf_token_valid(None, response['token']))
+
+    def test_session_begin_handler_is_not_rejected(self):
+        call_counter = self.exit_stack.enter_context(
+            self.swap_with_call_counter(
+                auth_services, 'establish_auth_session'))
+
+        self.get_html_response('/session_begin', expected_status_int=200)
+
+        self.assertEqual(call_counter.times_called, 1)
+
+    def test_session_end_handler_is_not_rejected(self):
+        call_counter = self.exit_stack.enter_context(
+            self.swap_with_call_counter(auth_services, 'destroy_auth_session'))
+
+        self.get_html_response('/session_end', expected_status_int=200)
+
+        self.assertEqual(call_counter.times_called, 1)
+
+    def test_seed_firebase_handler_is_not_rejected(self):
+        call_counter = self.exit_stack.enter_context(
+            self.swap_with_call_counter(auth_services, 'seed_firebase'))
+
+        response = (
+            self.get_html_response('/seed_firebase', expected_status_int=302))
+
+        self.assertEqual(call_counter.times_called, 1)
+        self.assertEqual(response.location, 'http://localhost/')
+
```

Comment:
Can you add backend integration tests to ensure that:

(a) users cannot create new accounts while in maintenance mode

(b) non-admin users are auto-logged-out if they try to log in

(c) any existing sessions are terminated if maintenance mode is switched on midway?

For (a) and (b), there should be exactly the same actions run in both maintenance and non-maintenance mode contexts (it's fine to have the "control group" case in this test suite too). The outcomes should be different.

Reply:
Done.

Making a stateful test will involve some non-intuitive changes (fetch `os.environ['EMAIL']` etc.; create JSON cookie string and place it in request; decode JSON cookie in `get_auth_claims_from_request`; have all test code call establish session somehow). I don't think it's worth complicating this PR, since the current test (check that `destroy_auth_session` is called) does a good job on its own.

Reply:
I've filed this as a FR in #12692 and have assigned myself.